sources/dynamodb/streaming.go

```diff
@@ -253,9 +316,119 @@ func getRecords(streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, shardItera
+	if eventName == "REMOVE" {
+		srcImage = record.Dynamodb.Keys
+	} else {
+		srcImage = record.Dynamodb.NewImage
+	}
+
+	spVals, badCols, srcStrVals := cvtRow(srcImage, srcSchema, spSchema, spCols)
+	if len(badCols) == 0 {
+		writeRecord(streamInfo, srcTable, spTable, eventName, spCols, spVals, srcSchema)
+	} else {
+		streamInfo.StatsAddBadRecord(srcTable, eventName)
+		streamInfo.CollectBadRecord(eventName, srcTable, srcSchema.ColNames, srcStrVals)
+	}
 	streamInfo.StatsAddRecordProcessed()
 }
+
+// writeRecord handles creation and processing of mutation from the converted data to Cloud Spanner.
+// If the writer which writes mutations to Cloud Spanner is not configured then it treats the record
+// as a bad record.
+func writeRecord(streamInfo *StreamingInfo, srcTable, spTable, eventName string, spCols []string, spVals []interface{}, srcSchema schema.Table) {
+	if streamInfo.write == nil {
+		msg := "Internal error: writeRecord called but writer not configured"
+		streamInfo.StatsAddBadRecord(srcTable, eventName)
+		streamInfo.Unexpected(msg)
+	} else {
+		m := getMutation(eventName, srcTable, spTable, spCols, spVals, srcSchema)
+		err := writeMutation(m, streamInfo)
+		if err != nil {
+			streamInfo.StatsAddDroppedRecord(srcTable, eventName)
+			streamInfo.CollectDroppedRecord(eventName, spTable, spCols, spVals, err)
+		}
+	}
+}
+
+// getMutation creates a mutation for writing to Cloud Spanner from the converted data.
+func getMutation(eventName, srcTable, spTable string, spCols []string, spVals []interface{}, srcSchema schema.Table) (m *sp.Mutation) {
+	if eventName == "INSERT" {
+		m = sp.Insert(spTable, spCols, spVals)
+	} else if eventName == "MODIFY" {
+		m = sp.InsertOrUpdate(spTable, spCols, spVals)
+	} else {
+		m = removeMutation(srcSchema, spTable, srcTable, spVals)
+	}
+	return m
+}
+
+// removeMutation create a mutation from converted data for records of type 'REMOVE'.
+// It ensures that when keyset is created the order for primary keys passed is same
+// as the original database i.e. HASH Key, Partition Key.
+func removeMutation(srcSchema schema.Table, spTable, srcTable string, spVals []interface{}) (m *sp.Mutation) {
+	var srcKeys []string
+	var reqSpVals []interface{}
+	for i := 0; i < len(spVals); i++ {
+		if spVals[i] == nil {
+			continue
+		}
+		srcKeys = append(srcKeys, srcSchema.ColNames[i])
+		reqSpVals = append(reqSpVals, spVals[i])
+	}
+	primaryKeys := srcSchema.PrimaryKeys
+	if primaryKeys[0].Column != srcKeys[0] {
+		reqSpVals[0], reqSpVals[1] = reqSpVals[1], reqSpVals[0]
+	}
+	if len(reqSpVals) == 1 {
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0]})
+	} else {
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0], reqSpVals[1]})
+	}
+	return m
+}
+
+// parentDataMissingError checks if the error is a parent data missing error.
+func parentDataMissingError(err error) bool {
+	return strings.Contains(err.Error(), "NotFound") && strings.Contains(err.Error(), "Parent row") && strings.Contains(err.Error(), "is missing")
```

Comment:
This is a pretty brittle piece of code. Depending on a the text of an error is risky. Unlike err types which are not usually change without notice to the client, the error text may change without notice.
If there is no way around this at this time, let's add a comment where on why we did this and create an issue in the repo to check we can come back and fix it later.

Reply:
As this error is very specific in this solution, adding a comment on why we are doing this and a note that this may be subject to change.